File: app/src/main/java/com/pitchai/flickrrecentphotobrowser/domain/JobExecutor.java
package com.pitchai.flickrrecentphotobrowser.domain;

import android.support.annotation.NonNull;


import com.pitchai.flickrrecentphotobrowser.domain.executor.ThreadExecutor;

import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;
import javax.inject.Singleton;

/**
 * Created by pitchairajamani on 2/22/17.
 */
//@Singleton
public class JobExecutor implements ThreadExecutor {
    private  final ThreadPoolExecutor threadPoolExecutor;

    //@Inject
    public JobExecutor() {
            this.threadPoolExecutor = new ThreadPoolExecutor(3, 5, 10, TimeUnit.SECONDS,
                    new LinkedBlockingQueue<Runnable>(), new JobThreadFactory());
    }

    @Override public void execute(@NonNull Runnable runnable) {
        this.threadPoolExecutor.execute(runnable);
    }

    private static class JobThreadFactory implements ThreadFactory {
        private int counter = 0;

        @Override public Thread newThread(@NonNull Runnable runnable) {
            return new Thread(runnable, "android_" + counter++);
        }
    }
}

File: app/src/main/java/com/pitchai/flickrrecentphotobrowser/domain/JobExecutor.kt
package com.pitchai.flickrrecentphotobrowser.domain

import com.pitchai.flickrrecentphotobrowser.domain.executor.ThreadExecutor
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.ThreadFactory
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

/**
 * Created by pitchairajamani on 2/22/17.
 */
//@Singleton
class JobExecutor : ThreadExecutor {
    private val threadPoolExecutor: ThreadPoolExecutor
    override fun execute(runnable: Runnable) {
        threadPoolExecutor.execute(runnable)
    }

    private class JobThreadFactory : ThreadFactory {
        private var counter = 0
        override fun newThread(runnable: Runnable): Thread {
            return Thread(runnable, "android_" + counter++)
        }
    }

    //@Inject
    init {
        threadPoolExecutor = ThreadPoolExecutor(3, 5, 10, TimeUnit.SECONDS,
                LinkedBlockingQueue(), JobThreadFactory())
    }
}